File: implementacion_java/src/algoritmos/SequentialBlockIV.java
package algoritmos;

import utilidades.AlgoritmoMultiplicacion;
import utilidades.Convertidor;

import java.math.BigInteger;

public class SequentialBlockIV implements AlgoritmoMultiplicacion {

    Convertidor convertidor = new Convertidor();

    @Override
    public void multiplicar(BigInteger[][] matriz1, BigInteger[][] matriz2) {
        // Convertir de BigInteger a int
        int[][] doubleMatriz1 = convertidor.convertBigIntegerArrayToIntArray(matriz1);
        int[][] doubleMatriz2 = convertidor.convertBigIntegerArrayToIntArray(matriz2);
        int[][] doubleResultado = sequentialBlock_IV3(doubleMatriz1, doubleMatriz2, doubleMatriz1.length);

        // Convertir el resultado int a BigINteger
        convertidor.convertIntArrayToBigInteger(doubleResultado);
    }

    public int[][] sequentialBlock_IV3(int[][] A, int[][] B, int block_size) {
        int n = A.length; // Tamaño de la matriz (se asume que es cuadrada)
        int[][] C = new int[n][n]; // Matriz de resultado

        // Recorre las matrices en bloques de tamaño block_size
        for (int i = 0; i < n; i += block_size) { // Recorrido por filas con incrementos de tamaño block_size
            for (int j = 0; j < n; j += block_size) { // Recorrido por columnas con incrementos de tamaño block_size
                for (int k = 0; k < n; k += block_size) { // Recorrido para la multiplicación de bloques
                    // Multiplica los bloques de matrices
                    for (int ii = i; ii < i + block_size && ii < n; ii++) { // Recorrido por filas dentro del bloque
                        for (int jj = j; jj < j + block_size && jj < n; jj++) { // Recorrido por columnas dentro del bloque
                            for (int kk = k; kk < k + block_size && kk < n; kk++) { // Recorrido para la multiplicación de elementos dentro del bloque
                                // Realiza la multiplicación del bloque actual y acumula el resultado en la matriz C
                                C[ii][kk] += A[ii][jj] * B[jj][kk];
                            }
                        }
                    }
                }
            }
        }

        return C; // Devolver la matriz de resultado
    }
}

File: implementacion_kotlin/src/multiplicacion_matrices/SequentialBlockIV.kt
package multiplicacion_matrices

import utilidades.AlgoritmoMultiplicacion
import utilidades.Convertidor
import java.math.BigInteger

class SequentialBlockIV : AlgoritmoMultiplicacion {
    var convertidor: Convertidor = Convertidor()

    override fun multiplicar(matriz1: Array<Array<BigInteger>>, matriz2: Array<Array<BigInteger>>) {
        // Convertir de BigInteger a int
        val doubleMatriz1 = convertidor.convertBigIntegerArrayToIntArray(matriz1)
        val doubleMatriz2 = convertidor.convertBigIntegerArrayToIntArray(matriz2)
        val doubleResultado = sequentialBlock_IV3(doubleMatriz1, doubleMatriz2, doubleMatriz1.size)

        // Convertir el resultado int a BigINteger
        convertidor.convertIntArrayToBigInteger(doubleResultado)
    }

    fun sequentialBlock_IV3(A: Array<IntArray>, B: Array<IntArray>, block_size: Int): Array<IntArray> {
        val n = A.size
        val C = Array(n) { IntArray(n) }

        // Recorre las matrices en bloques de tamaño block_size
        var i = 0
        while (i < n) {
            var j = 0
            while (j < n) {
                var k = 0
                while (k < n) {
                    // Multiplica los bloques de matrices
                    var ii = i
                    while (ii < i + block_size && ii < n) {
                        var jj = j
                        while (jj < j + block_size && ii < n) {
                            var kk = k
                            while (kk < k + block_size && k < n) {
                                C[ii][kk] += A[ii][jj] * B[jj][kk]
                                kk++
                            }
                            jj++
                        }
                        ii++
                    }
                    k += block_size
                }
                j += block_size
            }
            i += block_size
        }
        return C
    }

    // Método para convertir BigInteger[] a int[]
    private fun convertBigIntegerArrayToDoubleArray(bigIntArray: Array<Array<BigInteger>>): Array<IntArray> {
        val doubleArray = Array(bigIntArray.size) { IntArray(bigIntArray.size) }
        for (i in bigIntArray.indices) {
            for (j in bigIntArray.indices) {
                doubleArray[i][j] = bigIntArray[i][j].toInt()
            }
        }
        return doubleArray
    }

    // Método para convertir int[] a BigInteger
    private fun convertDoubleArrayToBigInteger(doubleArray: Array<IntArray>): Array<Array<BigInteger?>> {
        val rows = doubleArray.size
        val cols = doubleArray[0].size
        val bigIntegerArray = Array(rows) { arrayOfNulls<BigInteger>(cols) }

        for (i in 0 until rows) {
            for (j in 0 until cols) {
                // Convierte cada número decimal a BigInteger y almacénalo en la matriz
                // resultante
                bigIntegerArray[i][j] = BigInteger.valueOf(doubleArray[i][j].toLong())
            }
        }

        return bigIntegerArray
    }
}
